packages/api/src/command/medical/document/document-bulk-download.ts

```diff
@@ -0,0 +1,170 @@
+import { Patient } from "../../../domain/medical/patient";
+import { uuidv7 } from "@metriport/core/util/uuid-v7";
+import { storeBulkDownloadQueryInit } from "../patient/query-init";
+import { appendDocBulkDownloadProgress } from "../patient/append-bulk-doc-download-progress";
+import { DocumentReference } from "@medplum/fhirtypes";
+import { makeLambdaClient } from "../../../external/aws/lambda";
+import { getLambdaResultPayload } from "@metriport/core/external/aws/lambda";
+import {
+  DocumentBulkSignerLambdaRequest,
+  DocumentBulkSignerLambdaResponse,
+} from "@metriport/core/external/aws/lambda-logic/document-bulk-signing";
+
+const BATCH_SIZE = 100;
+const lambdaClient = makeLambdaClient();
+const bulkSigningLambdaName = Config.getBulkUrlSigningLambdaName();
+
+/**
+ * The function `triggerBulkUrlSigning` triggers the bulk signing process lambda for a patient's documents and
+ * returns the progress of the bulk download. It also triggers the webhook for the bulk download.
+ * @param {string} cxId - cxId
+ * @param {string} patientId - patientId
+ * @returns a Promise that resolves to a DocumentBulkDownloadProgress object.
+ */
+export const triggerBulkUrlSigning = async (
+  cxId: string,
+  patientId: string
+): Promise<DocumentBulkDownloadProgress> => {
+  if (!bulkSigningLambdaName) throw new Error("Bulk Signing Lambda Name is undefined");
+  const { log } = Util.out(`triggerBulkSigning - M patient ${patientId}`);
+
+  const patient = await getPatientOrFail({ id: patientId, cxId });
+
+  const docBulkDownloadProgress = patient.data.documentBulkDownloadProgress;
+  const requestId = getOrGenerateRequestId(docBulkDownloadProgress);
+
+  if (isDocBulkDownloadProcessing(docBulkDownloadProgress?.download)) {
+    log(
+      `Patient ${patientId}, Request ${requestId}, docBulkDownloadProgress is already 'processing', skipping...`
+    );
+    return createBulkDownloadQueryResponse("processing", patient);
+  }
+
+  const documents = await searchDocuments({ cxId, patientId });
+
+  let updatedPatient = await storeBulkDownloadQueryInit({
+    id: patient.id,
+    cxId: patient.cxId,
+    documentBulkDownloadProgress: { download: { status: "processing" } },
+    requestId,
+    totalDocuments: documents.length,
+  });
+
+  let successes = 0;
+  const errors = 0;
+  // Process each batch
+  const batches = chunk(documents, BATCH_SIZE);
+  for (let batchIndex = 0; batchIndex < batches.length; batchIndex++) {
+    const batch = batches[batchIndex] as DocumentReference[];
+    let dtos: DocumentBulkDownloadDTO[] = [];
+
+    const payload: DocumentBulkSignerLambdaRequest = {
+      patientId: patientId,
+      cxId: cxId,
+      documents: batch,
+    };
+
+    //Invoke the lambda function
+    const result = await lambdaClient
+      .invoke({
+        FunctionName: bulkSigningLambdaName,
+        InvocationType: "RequestResponse",
+        Payload: JSON.stringify(payload),
+      })
+      .promise();
```

Comment:
We're directly invoking the lambda, which in theory is just adding more time to the overall processing of the original request. If we take this approach, we might as well just presign the URLs on the API itself, IMO.

I suggested that we send these as SQS messages, so we essentially offload the processing to the lambda; when it finishes processing each batch, it calls an internal endpoint on the API indicating its done.

The main reason for this is so the time it takes for the API to process each request is limited to a few seconds - instead of a long running process that could be broken if we make a release midway through it.

Reply:
I would also consider adding a sweeper function/endpoint + scheduled lambda that calls it, assuming we take the offload approach: there's non-zero chance of at least one batch failing @ lambda and we never get the last message, in which case we would prob want to force-finish the process after a certain amount of time.